cdhweb/people/models.py

```diff
@@ -371,70 +381,115 @@ def current_title(self):
+    subpage_types = [ProfilePage]
+
+
 class Position(DateRange):
     '''Through model for many-to-many relation between people
     and titles.  Adds start and end dates to the join table.'''
-    user = models.ForeignKey(User, on_delete=models.CASCADE,
-                             related_name='positions')
+    person = ParentalKey(Person, on_delete=models.CASCADE,
+                         related_name="positions")
     title = models.ForeignKey(Title, on_delete=models.CASCADE)
 
     class Meta:
         ordering = ['-start_date']
 
     def __str__(self):
-        return '%s %s (%s)' % (self.user, self.title, self.start_date.year)
+        return '%s %s (%s)' % (self.person, self.title, self.start_date.year)
 
 
-def init_profile_from_ldap(user, ldapinfo):
-    '''Extra user/profile init logic for auto-populating people
-    profile fields with data available in LDAP.'''
+def init_person_from_ldap(user, ldapinfo):
+    '''Extra User init logic for creating and auto-populating a corresponding
+    Person with data from LDAP.'''
 
+    # update User to use ldap email
     user.email = user.email.lower()
     user.save()
 
-    try:
-        profile = user.profile
-    except ObjectDoesNotExist:
-        profile = Profile.objects.create(user=user)
-        # set profiles to draft by default when creating a new profile *only
-        # so we don't get a new page for every account we initialize
-        profile.status = CONTENT_STATUS_DRAFT
-
-    # populate profile with data we can pull from ldap
-    # but do not set any fields with content, which may contain edits
-
-    # - set user's display name as page title
-    if not profile.title:
-        profile.title = str(ldapinfo.displayName)
-    # - generate a slug based on display name
-    if not profile.slug:
-        profile.slug = slugify(ldapinfo.displayName)
-    if ldapinfo.telephoneNumber and not profile.phone_number:
-        profile.phone_number = str(ldapinfo.telephoneNumber)
-    # 'street' in ldap is office location
-    if ldapinfo.street and not profile.office_location:
-        profile.office_location = str(ldapinfo.street)
-    # organizational unit = department
-    if ldapinfo.ou and not profile.department:
-        profile.department = str(ldapinfo.ou)
-    # Store job title as string.
-    # NOTE: we may want to split title and only use the first portion.
-    # if ldapinfo.title and not profile.job_title:
-    profile.job_title = str(ldapinfo.title).split('.')[0]
+    # populate Person fields with data we can pull from ldap, if they are empty
+    person, _created = Person.objects.get_or_create(user=user)
 
-    # always update PU status to current
-    profile.pu_status = str(ldapinfo.pustatus)
+    # set first/last name basic on basic whitespace split; can adjust later
+    first_name, *_others, last_name = str(ldapinfo.displayName).split()
+    person.first_name = person.first_name or first_name
+    person.last_name = person.last_name or last_name
```

Comment:
Nice syntax trick, but I don't think we should be splitting displayName here.

We have an attribute map configured in settings that pucas picks up, which sets first name from `givenName` and last name from `sn` (surname). I think the name fields should be copied from user to person instead, if not already set.

Reply:
how would i access `first_name` and `last_name` in this context? i see the settings but not sure how it works.

re: copying from user to person, wouldn't this already happen as a result of the migrations? i don't think that user has a `first_name` or `last_name` after that, because it's just a regular django user model .

Reply:
Copying from user to person only happens in migrations for existing records; this method is intended primarily for populating new records.

The user model has existing `first_name` and `last_name` properties that are part of the regular django user model (see https://docs.djangoproject.com/en/3.1/ref/contrib/auth/#fields), which will already be populated from LDAP. It makes more sense to me to use those fields rather than getting them directly from ldapinfo, since if the fields we need change we want to be able to change in just one place.

Reply:
didn't realize that django's user model had these fields too! makes sense to just copy them on `Person` init then. does it matter that we would have:
- `User.first_name` / `User.last_name`
- `Person.first_name` / `Person.last_name`
- `Profile.title`

which can all be different? I guess nobody's really going to change the ones on `User`.

Reply:
Since `Person.user` is optional, we need to duplicate the fields. I think the names on `user` are mostly for convenience, so we know which user accounts we're managing. It might be nice to sync them from `person` to `user` on save, but not sure it's worth the hassle.

Is `title` required on profile page? The title field was required for mezzanine, and I think it was sometimes a point of confusion where you needed to edit that. It would be nice if we could get rid of that one, so it wouldn't be possible for the profile page title to be different from the person's name. (I wonder if we have any cases where they differ now?)

Reply:
It is required for `Page` subtypes. I actually implemented exactly that, until I realized that it differs for a decent number of people - e.g. grant, who is `Grant R. Wythoff` in his profile. So unfortunately I don't think we can do it unless we want to take away those middle initials.

Reply:
I noticed the middle initial on his a while back and wondered if he wanted it there but never asked. I _think_ this is based on a difference between the first and last names and display name in LDAP. But we might still want the flexibility to have a full display name that _can_ be different from first and last...